File: src/test/java/mission/step3/MenuServiceTest.java
package mission.step3;

import kitchenpos.application.MenuService;
import kitchenpos.domain.*;
import kitchenpos.infra.PurgomalumClient;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;

@ExtendWith(MockitoExtension.class)
class MenuServiceTest {

    @Mock
    private MenuRepository menuRepository;
    @Mock
    private MenuGroupRepository menuGroupRepository;
    @Mock
    private ProductRepository productRepository;
    @Mock
    private PurgomalumClient purgomalumClient;

    @InjectMocks
    private MenuService menuService;

    @Nested
    @DisplayName("메뉴 생성")
    class CreateMenu {

        @Test
        @DisplayName("메뉴를 생성한다")
        void create() {
            // given
            UUID productId = UUID.randomUUID();
            UUID menuGroupId = UUID.randomUUID();

            MenuGroup menuGroup = new MenuGroup();
            menuGroup.setId(menuGroupId);
            menuGroup.setName("음료");

            Product product = new Product();
            product.setId(productId);
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProduct(product);
            menuProduct.setProductId(productId);
            menuProduct.setQuantity(1);

            Menu request = new Menu();
            request.setName("아메리카노");
            request.setPrice(BigDecimal.valueOf(4000));
            request.setMenuGroupId(menuGroupId);
            request.setDisplayed(true);
            request.setMenuProducts(List.of(menuProduct));

            given(menuGroupRepository.findById(menuGroupId))
                    .willReturn(Optional.of(menuGroup));
            given(productRepository.findAllByIdIn(List.of(productId)))
                    .willReturn(List.of(product));
            given(productRepository.findById(productId))
                    .willReturn(Optional.of(product));
            given(purgomalumClient.containsProfanity(request.getName()))
                    .willReturn(false);
            given(menuRepository.save(any(Menu.class)))
                    .willAnswer(invocation -> {
                        Menu saved = invocation.getArgument(0);
                        saved.setId(UUID.randomUUID());
                        return saved;
                    });

            // when
            Menu created = menuService.create(request);

            // then
            assertThat(created.getId()).isNotNull();
            assertThat(created.getName()).isEqualTo("아메리카노");
            assertThat(created.getPrice()).isEqualByComparingTo(BigDecimal.valueOf(4000));
            assertThat(created.getMenuGroup()).isEqualTo(menuGroup);
            assertThat(created.isDisplayed()).isTrue();
            assertThat(created.getMenuProducts()).hasSize(1);
        }

        @Test
        @DisplayName("가격이 메뉴 상품의 가격 합계보다 크면 예외가 발생한다")
        void createWithPriceGreaterThanTotalProductPrice() {
            // given
            UUID menuGroupId = UUID.randomUUID();
            UUID productId = UUID.randomUUID();

            MenuGroup menuGroup = new MenuGroup();
            menuGroup.setId(menuGroupId);

            Product product = new Product();
            product.setId(productId);
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProductId(productId);
            menuProduct.setQuantity(1);

            Menu request = new Menu();
            request.setName("아메리카노");
            request.setPrice(BigDecimal.valueOf(5000));
            request.setMenuGroupId(menuGroupId);
            request.setMenuProducts(List.of(menuProduct));

            given(menuGroupRepository.findById(menuGroupId))
                    .willReturn(Optional.of(menuGroup));
            given(productRepository.findAllByIdIn(List.of(productId)))
                    .willReturn(List.of(product));
            given(productRepository.findById(productId))
                    .willReturn(Optional.of(product));

            // when & then
            assertThatThrownBy(() -> menuService.create(request))
                    .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        @DisplayName("메뉴 그룹이 존재하지 않으면 예외가 발생한다")
        void createWithNonExistentMenuGroup() {
            // given
            Menu request = new Menu();
            request.setName("아메리카노");
            request.setPrice(BigDecimal.valueOf(4000));
            request.setMenuGroupId(UUID.randomUUID());

            given(menuGroupRepository.findById(any()))
                    .willReturn(Optional.empty());

            // when & then
            assertThatThrownBy(() -> menuService.create(request))
                    .isInstanceOf(NoSuchElementException.class);
        }

        @Test
        @DisplayName("메뉴 상품의 수량이 0 이하면 예외가 발생한다")
        void createWithInvalidProductQuantity() {
            // given
            UUID menuGroupId = UUID.randomUUID();
            UUID productId = UUID.randomUUID();

            MenuGroup menuGroup = new MenuGroup();
            menuGroup.setId(menuGroupId);

            Product product = new Product();
            product.setId(productId);
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProductId(productId);
            menuProduct.setQuantity(0);

            Menu request = new Menu();
            request.setName("아메리카노");
            request.setPrice(BigDecimal.valueOf(4000));
            request.setMenuGroupId(menuGroupId);
            request.setMenuProducts(List.of(menuProduct));

            given(menuGroupRepository.findById(menuGroupId))
                    .willReturn(Optional.of(menuGroup));
            given(productRepository.findAllByIdIn(List.of(productId)))
                    .willReturn(List.of(product));
            given(productRepository.findById(productId))
                    .willReturn(Optional.of(product));

            // when & then
            assertThatThrownBy(() -> menuService.create(request))
                    .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    @DisplayName("메뉴 가격 변경")
    class ChangePrice {
        @Test
        @DisplayName("메뉴의 가격을 변경한다")
        void changePrice() {
            // given
            UUID menuId = UUID.randomUUID();
            Product product = new Product();
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProduct(product);
            menuProduct.setQuantity(1);

            Menu menu = new Menu();
            menu.setId(menuId);
            menu.setPrice(BigDecimal.valueOf(4000));
            menu.setMenuProducts(List.of(menuProduct));

            Menu request = new Menu();
            request.setPrice(BigDecimal.valueOf(4000));

            given(menuRepository.findById(menuId))
                    .willReturn(Optional.of(menu));

            // when
            Menu updated = menuService.changePrice(menuId, request);

            // then
            assertThat(updated.getPrice()).isEqualByComparingTo(BigDecimal.valueOf(4000));
        }
    }

    @Nested
    @DisplayName("메뉴 표시 상태 변경")
    class DisplayStatus {
        @Test
        @DisplayName("메뉴를 보임 상태로 변경한다")
        void display() {
            // given
            UUID menuId = UUID.randomUUID();
            Product product = new Product();
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProduct(product);
            menuProduct.setQuantity(1);

            Menu menu = new Menu();
            menu.setId(menuId);
            menu.setPrice(BigDecimal.valueOf(4000));
            menu.setMenuProducts(List.of(menuProduct));
            menu.setDisplayed(false);

            given(menuRepository.findById(menuId))
                    .willReturn(Optional.of(menu));

            // when
            Menu displayed = menuService.display(menuId);

            // then
            assertThat(displayed.isDisplayed()).isTrue();
        }

        @Test
        @DisplayName("메뉴 가격이 구성 상품들의 총합보다 높으면 보임 상태로 변경되지 않고 예외가 발생한다")
        void displayFailsWhenMenuPriceExceedsTotalProductPrice() {

            // given
            UUID menuId = UUID.randomUUID();
            Product product = new Product();
            product.setPrice(BigDecimal.valueOf(4000));

            MenuProduct menuProduct = new MenuProduct();
            menuProduct.setProduct(product);
            menuProduct.setQuantity(1);

            Menu menu = new Menu();
            menu.setId(menuId);

            menu.setPrice(BigDecimal.valueOf(5000));
            menu.setMenuProducts(List.of(menuProduct));
            menu.setDisplayed(false);

            given(menuRepository.findById(menuId))
                    .willReturn(Optional.of(menu));

            // when & then
            assertThatThrownBy(() -> menuService.display(menuId))
                    .isInstanceOf(IllegalStateException.class);
        }

        @Test
        @DisplayName("메뉴를 숨김 상태로 변경한다")
        void hide() {
            // given
            UUID menuId = UUID.randomUUID();
            Menu menu = new Menu();
            menu.setId(menuId);
            menu.setDisplayed(true);

            given(menuRepository.findById(menuId))
                    .willReturn(Optional.of(menu));

            // when
            Menu hidden = menuService.hide(menuId);

            // then
            assertThat(hidden.isDisplayed()).isFalse();
        }
    }

    @Test
    @DisplayName("전체 메뉴를 조회한다")
    void findAll() {
        // given
        Menu menu1 = new Menu();
        menu1.setId(UUID.randomUUID());
        menu1.setName("아메리카노");
        menu1.setPrice(BigDecimal.valueOf(4000));

        Menu menu2 = new Menu();
        menu2.setId(UUID.randomUUID());
        menu2.setName("카페라떼");
        menu2.setPrice(BigDecimal.valueOf(4500));

        given(menuRepository.findAll())
                .willReturn(List.of(menu1, menu2));

        // when
        List<Menu> menus = menuService.findAll();

        // then
        assertThat(menus).hasSize(2);
        assertThat(menus.get(0).getName()).isEqualTo("아메리카노");
        assertThat(menus.get(1).getName()).isEqualTo("카페라떼");
    }
}

File: src/test/kotlin/kitchenops/application/MenuServiceTest.kt
package kitchenops.application

import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kitchenpos.application.MenuService
import kitchenpos.domain.*
import kitchenpos.infra.PurgomalumClient
import spec.BaseUnitSpec
import java.math.BigDecimal
import java.util.*

internal class MenuServiceTest : BaseUnitSpec({

    val menuRepository = mockk<MenuRepository>()
    val menuGroupRepository = mockk<MenuGroupRepository>()
    val productRepository = mockk<ProductRepository>()
    val purgomalumClient = mockk<PurgomalumClient>()

    val sut = MenuService(menuRepository, menuGroupRepository, productRepository, purgomalumClient)

    context("메뉴를 생성할 수 있다") {
        test("가격을 지정하지 않으면 생성할 수 없다") {
            // given
            val request = createMenu(price = null)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("가격이 0원 미만이면 생성할 수 없다") {
            // given
            val request = createMenu(price = BigDecimal("-1"))

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("메뉴 그룹이 존재하지 않으면 생성할 수 없다") {
            // given
            val menuGroup = createMenuGroup(id = UUID.randomUUID())
            val request = createMenu(price = BigDecimal("0"), menuGroup = menuGroup)

            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.empty()

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe NoSuchElementException()
        }

        test("사용하는 상품을 지정하지 않으면 생성할 수 없다") {
            // given
            val menuGroup = createMenuGroup(id = UUID.randomUUID())
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request = createMenu(price = BigDecimal("0"), menuProducts = null, menuGroup = menuGroup)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("사용하는 상품이 비어있으면 생성할 수 없다") {
            // given
            val menuGroup = createMenuGroup(id = UUID.randomUUID())
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request = createMenu(price = BigDecimal("100"), menuProducts = emptyList(), menuGroup = menuGroup)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("사용하는 상품이 존재하지 않으면 생성할 수 없다") {
            // given
            val invalidProduct = createProduct(id = UUID.randomUUID())
            val menuProduct = createMenuProduct(product = invalidProduct, quantity = 1)
            every { productRepository.findAllByIdIn(listOf(invalidProduct.id)) } returns emptyList()

            val menuGroup = MenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request =
                createMenu(price = BigDecimal("100"), menuProducts = listOf(menuProduct), menuGroup = menuGroup)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("사용하는 상품의 수량이 0개 미만이면 생성할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 0, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns emptyList()

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request =
                createMenu(price = BigDecimal("100"), menuProducts = listOf(menuProduct), menuGroup = menuGroup)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("메뉴의 가격이 상품 가격의 총 합보다 클 경우 생성할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 3, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns listOf(product)

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request =
                createMenu(price = BigDecimal("500"), menuProducts = listOf(menuProduct), menuGroup = menuGroup)

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("이름을 지정하지 않으면 생성할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 1, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns listOf(product)

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request = createMenu(
                price = BigDecimal("100"),
                menuProducts = listOf(menuProduct),
                menuGroup = menuGroup,
                name = null
            )

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        // TODO: empty name에 동작하지 않아야 할 것 같은데 버그인 듯 하다
        test("이름이 비어있으면 생성할 수 없다").config(enabled = false) {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 1, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns listOf(product)

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request = createMenu(
                price = BigDecimal("100"),
                menuProducts = listOf(menuProduct),
                menuGroup = menuGroup,
                name = "",
            )

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("메뉴의 이름에 욕설이 포함되면 생성할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 1, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns listOf(product)

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            val request = createMenu(
                price = BigDecimal("100"),
                menuProducts = listOf(menuProduct),
                menuGroup = menuGroup,
                name = "hell",
            )

            every { purgomalumClient.containsProfanity("hell") } returns true

            // when
            val actual = runCatching { sut.create(request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("생성한다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            every { productRepository.findById(product.id) } returns Optional.of(product)

            val menuProduct = createMenuProduct(quantity = 1, product = product)
            every { productRepository.findAllByIdIn(listOf(menuProduct.productId)) } returns listOf(product)

            val menuGroup = createMenuGroup()
            every { menuGroupRepository.findById(menuGroup.id) } returns Optional.of(menuGroup)

            every { purgomalumClient.containsProfanity("myName") } returns false

            every { menuRepository.save(any()) } returns createMenu(
                price = BigDecimal("100"),
                menuProducts = listOf(menuProduct),
                menuGroup = menuGroup,
                name = "myName",
                displayed = true
            )

            val request = createMenu(
                price = BigDecimal("100"),
                menuProducts = listOf(menuProduct),
                menuGroup = menuGroup,
                name = "myName",
                displayed = true,
            )

            // when
            val actual = sut.create(request)

            // then
            actual should {
                it.name shouldBe "myName"
                it.price shouldBe BigDecimal("100")
                it.isDisplayed shouldBe true
                it.menuGroup shouldBe menuGroup
                it.menuGroupId shouldBe menuGroup.id
                it.menuProducts shouldBe listOf(menuProduct)
            }

            verify(exactly = 1) { menuRepository.save(any()) }
        }
    }

    context("메뉴를 전체 조회할 수 있다") {
        test("전체 조회한다") {
            // given
            val menu1 = createMenu()
            val menu2 = createMenu()
            every { menuRepository.findAll() } returns listOf(menu1, menu2)

            // when
            val actual = sut.findAll()

            // then
            actual shouldBe listOf(menu1, menu2)
        }
    }

    context("메뉴의 가격을 변경할 수 있다") {
        test("가격을 지정하지 않으면 변경할 수 없다") {
            // given
            val request = createMenu(price = null)

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("가격이 0원 미만이면 변경할 수 없다") {
            // given
            val request = createMenu(price = BigDecimal("-1"))

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("메뉴가 존재하지 않으면 변경할 수 없다") {
            // given
            val request = createMenu(price = BigDecimal("100"))

            every { menuRepository.findById(request.id) } returns Optional.empty()

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe NoSuchElementException()
        }

        test("가격이 상품 가격의 합보다 크면 생성할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            val menuProduct = createMenuProduct(quantity = 3, product = product)
            val menu = createMenu(menuProducts = listOf(menuProduct))

            every { menuRepository.findById(menu.id) } returns Optional.of(menu)

            val request = createMenu(id = menu.id, price = BigDecimal("400"))

            // when
            val actual = runCatching { sut.changePrice(request.id, request) }

            // then
            actual.exceptionOrNull() shouldBe IllegalArgumentException()
        }

        test("가격을 변경한다") {
            // given
            val product = createProduct(price = BigDecimal("1000"))
            val menuProduct = createMenuProduct(quantity = 1, product = product)
            val menu = createMenu(menuProducts = listOf(menuProduct))

            every { menuRepository.findById(menu.id) } returns Optional.of(menu)

            val request = createMenu(id = menu.id, price = BigDecimal("1000"))

            // when
            val actual = sut.changePrice(request.id, request)

            // then
            actual.price shouldBe BigDecimal("1000")
        }
    }

    context("메뉴를 표시하게 변경할 수 있다") {
        test("메뉴가 존재하지 않으면 변경할 수 없다") {
            // given
            val menuId = UUID.randomUUID()
            every { menuRepository.findById(menuId) } returns Optional.empty()

            // when
            val actual = runCatching { sut.display(menuId) }

            // then
            actual.exceptionOrNull() shouldBe NoSuchElementException()
        }

        test("메뉴의 가격이 상품 가격의 합보다 클 경우에는 화면에 표시할 수 없다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            val menuProduct = createMenuProduct(quantity = 1, product = product)

            val menu = createMenu(price = BigDecimal("1000"), menuProducts = listOf(menuProduct))
            every { menuRepository.findById(menu.id) } returns Optional.of(menu)

            // when
            val actual = runCatching { sut.display(menu.id) }

            // then
            actual.exceptionOrNull() shouldBe IllegalStateException()
        }

        test("표시하게 변경한다") {
            // given
            val product = createProduct(price = BigDecimal("100"))
            val menuProduct = createMenuProduct(quantity = 1, product = product)

            val menu = createMenu(price = BigDecimal("100"), menuProducts = listOf(menuProduct), displayed = false)
            every { menuRepository.findById(menu.id) } returns Optional.of(menu)

            // when
            val actual = sut.display(menu.id)

            // then
            actual.isDisplayed shouldBe true
        }
    }

    context("메뉴를 숨기게 변경할 수 있다") {
        test("메뉴가 존재하지 않으면 변경할 수 없다") {
            // given
            val menuId = UUID.randomUUID()
            every { menuRepository.findById(menuId) } returns Optional.empty()

            // when
            val actual = runCatching { sut.hide(menuId) }

            // then
            actual.exceptionOrNull() shouldBe NoSuchElementException()
        }

        test("숨기게 변경한다") {
            // given
            val menu = createMenu(displayed = true)
            every { menuRepository.findById(menu.id) } returns Optional.of(menu)

            // when
            val actual = sut.hide(menu.id)

            // then
            actual.isDisplayed shouldBe false
        }
    }
}) {
    companion object {
        fun createMenu(
            id: UUID? = null,
            name: String? = null,
            price: BigDecimal? = null,
            menuGroup: MenuGroup? = null,
            displayed: Boolean = true,
            menuProducts: List<MenuProduct>? = null
        ) = Menu().apply {
            this.id = id
            this.name = name
            this.price = price
            this.menuGroup = menuGroup
            this.menuGroupId = menuGroup?.id
            this.isDisplayed = displayed
            this.menuProducts = menuProducts
        }

        fun createMenuGroup(id: UUID? = null) = MenuGroup().apply {
            this.id = id
        }

        fun createMenuProduct(product: Product?, quantity: Long = 0) = MenuProduct().apply {
            this.product = product
            this.productId = product?.id
            this.quantity = quantity
        }

        fun createProduct(id: UUID? = null, price: BigDecimal? = null) = Product().apply {
            this.id = id
            this.price = price
        }
    }
}
